westend fauced stopped working because of rpc timeouts

facet no lognger woking:
```
faucet
10:59
An unexpected error occured, please check the server logs
```

output from metrics exporter:
```
/bot # wget -O - http://localhost:5555/metrics
Connecting to localhost:5555 (127.0.0.1:5555)
writing to stdout
# HELP errors_total The total amount of errors logged on the faucet backend
# TYPE errors_total counter
errors_total 33
# HELP errors_rpc_timeout The total amount of timeout errors between the faucet backend and the rpc node
# TYPE errors_rpc_timeout counter
errors_rpc_timeout 27
```

after restarting the pod (containing bot & server) system worked as expected again. the rpc node was available at all times (from what i can tell) what makes me think there might be a problem with timeout / connection handling inside the server app?

here the bot & server logs:
```
/bot # wget -O - http://localhost:5555/metrics
Connecting to localhost:5555 (127.0.0.1:5555)
writing to stdout
# HELP errors_total The total amount of errors logged on the faucet backend
# TYPE errors_total counter
errors_total 33
# HELP errors_rpc_timeout The total amount of timeout errors between the faucet backend and the rpc node
# TYPE errors_rpc_timeout counter
errors_rpc_timeout 27


bot log:
---
[...]
toJSON: [Function: toJSON]
}
Event sent to !cJFtAIkwxuofiSYkPN:matrix.org with event id $Id6NcKtGSd8_gZow2vVM3FZwDsGicGywPap-DSC_9Sw
setting pendingEvent status to sent in !cJFtAIkwxuofiSYkPN:matrix.org event ID ~!cJFtAIkwxuofiSYkPN:matrix.org:m1611826410783.346 -> $Id6NcKtGSd8_gZow2vVM3FZwDsGicGywPap-DSC_9Sw
Already have event $Id6NcKtGSd8_gZow2vVM3FZwDsGicGywPap-DSC_9Sw in limited sync - not resetting
Got remote echo for event $Id6NcKtGSd8_gZow2vVM3FZwDsGicGywPap-DSC_9Sw -> $Id6NcKtGSd8_gZow2vVM3FZwDsGicGywPap-DSC_9Sw old status sent
sendEvent of type m.room.message in !cJFtAIkwxuofiSYkPN:matrix.org with txnId m1611826421264.347
Event sent to !cJFtAIkwxuofiSYkPN:matrix.org with event id $EIyjiUmI4OfQSTjzrl5a0nJmVAN02YTcTsuOmocHS2k
setting pendingEvent status to sent in !cJFtAIkwxuofiSYkPN:matrix.org event ID ~!cJFtAIkwxuofiSYkPN:matrix.org:m1611826421264.347 -> $EIyjiUmI4OfQSTjzrl5a0nJmVAN02YTcTsuOmocHS2k
Already have event $EIyjiUmI4OfQSTjzrl5a0nJmVAN02YTcTsuOmocHS2k in limited sync - not resetting
Got remote echo for event $EIyjiUmI4OfQSTjzrl5a0nJmVAN02YTcTsuOmocHS2k -> $EIyjiUmI4OfQSTjzrl5a0nJmVAN02YTcTsuOmocHS2k old status sent
sendEvent of type m.room.message in !cJFtAIkwxuofiSYkPN:matrix.org with txnId m1611826761285.348
Event sent to !cJFtAIkwxuofiSYkPN:matrix.org with event id $zSSGvJ3xoANRqP7Vxe7-JmyyNAvOGuzfRGRwoCdmoQ4
setting pendingEvent status to sent in !cJFtAIkwxuofiSYkPN:matrix.org event ID ~!cJFtAIkwxuofiSYkPN:matrix.org:m1611826761285.348 -> $zSSGvJ3xoANRqP7Vxe7-JmyyNAvOGuzfRGRwoCdmoQ4
Already have event $zSSGvJ3xoANRqP7Vxe7-JmyyNAvOGuzfRGRwoCdmoQ4 in limited sync - not resetting
Got remote echo for event $zSSGvJ3xoANRqP7Vxe7-JmyyNAvOGuzfRGRwoCdmoQ4 -> $zSSGvJ3xoANRqP7Vxe7-JmyyNAvOGuzfRGRwoCdmoQ4 old status sent
sendEvent of type m.room.message in !cJFtAIkwxuofiSYkPN:matrix.org with txnId m1611827646167.349
[2021-01-28T09:54:06.168] [ERROR] default - An error occured when dripping Error: timeout of 10000ms exceeded
    at createError (/bot/node_modules/axios/lib/core/createError.js:16:15)
    at RedirectableRequest.handleRequestTimeout (/bot/node_modules/axios/lib/adapters/http.js:280:16)
    at Object.onceWrapper (node:events:482:28)
    at RedirectableRequest.emit (node:events:376:20)
    at Timeout._onTimeout (/bot/node_modules/follow-redirects/index.js:166:13)
    at listOnTimeout (node:internal/timers:556:17)
    at processTimers (node:internal/timers:499:7) {
  config: {
    url: '/bot-endpoint',
    method: 'post',
    data: '{"address":"5EuZQ2ZHEWBC4Dbu7uTBSaV7sqt3BxoNSzh1uU8vbnkBx65V","amount":1,"sender":"@uu3:matrix.org"}',
    headers: {
      Accept: 'application/json, text/plain, */*',
      'Content-Type': 'application/json;charset=utf-8',
      'User-Agent': 'axios/0.21.1',
      'Content-Length': 100
    },
    baseURL: 'http://localhost:5555',
    transformRequest: [ [Function: transformRequest] ],
    transformResponse: [ [Function: transformResponse] ],
    timeout: 10000,
    adapter: [Function: httpAdapter],
    xsrfCookieName: 'XSRF-TOKEN',
    xsrfHeaderName: 'X-XSRF-TOKEN',
    maxContentLength: -1,
    maxBodyLength: -1,
    validateStatus: [Function: validateStatus]
  },
  code: 'ECONNABORTED',
  request: <ref *1> Writable {
    _writableState: WritableState {
      objectMode: false,
      highWaterMark: 16384,
      finalCalled: false,
      needDrain: false,
      ending: false,
      ended: false,
      finished: false,
      destroyed: false,
      decodeStrings: true,
      defaultEncoding: 'utf8',
      length: 0,
      writing: false,
      corked: 0,
      sync: true,
      bufferProcessing: false,
      onwrite: [Function: bound onwrite],
      writecb: null,
      writelen: 0,
      afterWriteTickInfo: null,
      buffered: [],
      bufferedIndex: 0,
      allBuffers: true,
      allNoop: true,
      pendingcb: 0,
      constructed: true,
      prefinished: false,
      errorEmitted: false,
      emitClose: true,
      autoDestroy: true,
      errored: null,
      closed: false,
      closeEmitted: false,
      [Symbol(kOnFinished)]: []
    },
    _events: [Object: null prototype] { response: [Array], error: [Array] },
    _eventsCount: 2,
    _maxListeners: undefined,
    _options: {
      maxRedirects: 21,
      maxBodyLength: 10485760,
      protocol: 'http:',
      path: '/bot-endpoint',
      method: 'POST',
      headers: [Object],
      agent: undefined,
      agents: [Object],
      auth: undefined,
      hostname: 'localhost',
      port: '5555',
      nativeProtocols: [Object],
      pathname: '/bot-endpoint'
    },
    _ended: true,
    _ending: true,
    _redirectCount: 0,
    _redirects: [],
    _requestBodyLength: 100,
    _requestBodyBuffers: [ [Object] ],
    _onNativeResponse: [Function (anonymous)],
    _currentRequest: ClientRequest {
      _events: [Object: null prototype],
      _eventsCount: 7,
      _maxListeners: undefined,
      outputData: [],
      outputSize: 0,
      writable: true,
      destroyed: true,
      _last: true,
      chunkedEncoding: false,
      shouldKeepAlive: false,
      _defaultKeepAlive: true,
      useChunkedEncodingByDefault: true,
      sendDate: false,
      _removedConnection: false,
      _removedContLen: false,
      _removedTE: false,
      _contentLength: null,
      _hasBody: true,
      _trailer: '',
      finished: true,
      _headerSent: true,
      _closed: false,
      socket: [Socket],
      _header: 'POST /bot-endpoint HTTP/1.1\r\n' +
        'Accept: application/json, text/plain, */*\r\n' +
        'Content-Type: application/json;charset=utf-8\r\n' +
        'User-Agent: axios/0.21.1\r\n' +
        'Content-Length: 100\r\n' +
        'Host: localhost:5555\r\n' +
        'Connection: close\r\n' +
        '\r\n',
      _keepAliveTimeout: 0,
      _onPendingData: [Function: noopPendingOutput],
      agent: [Agent],
      socketPath: undefined,
      method: 'POST',
      maxHeaderSize: undefined,
      insecureHTTPParser: undefined,
      path: '/bot-endpoint',
      _ended: false,
      res: null,
      aborted: true,
      timeoutCb: null,
      upgradeOrConnect: false,
      parser: [HTTPParser],
      maxHeadersCount: null,
      reusedSocket: false,
      host: 'localhost',
      protocol: 'http:',
      _redirectable: [Circular *1],
      [Symbol(kCapture)]: false,
      [Symbol(kNeedDrain)]: false,
      [Symbol(corked)]: 0,
      [Symbol(kOutHeaders)]: [Object: null prototype]
    },
    _currentUrl: 'http://localhost:5555/bot-endpoint',
    _timeout: Timeout {
      _idleTimeout: 10000,
      _idlePrev: null,
      _idleNext: null,
      _idleStart: 667261484,
      _onTimeout: [Function (anonymous)],
      _timerArgs: undefined,
      _repeat: null,
      _destroyed: true,
      [Symbol(refed)]: true,
      [Symbol(kHasPrimitive)]: false,
      [Symbol(asyncId)]: 1577197,
      [Symbol(triggerId)]: 1577191
    },
    [Symbol(kCapture)]: false
  },
  response: undefined,
  isAxiosError: true,
  toJSON: [Function: toJSON]
}
Event sent to !cJFtAIkwxuofiSYkPN:matrix.org with event id $jFY2FyWh5xaypsaPLPYZ11AtDTa8zTPc5KF9SUUE6JY
setting pendingEvent status to sent in !cJFtAIkwxuofiSYkPN:matrix.org event ID ~!cJFtAIkwxuofiSYkPN:matrix.org:m1611827646167.349 -> $jFY2FyWh5xaypsaPLPYZ11AtDTa8zTPc5KF9SUUE6JY
Already have event $jFY2FyWh5xaypsaPLPYZ11AtDTa8zTPc5KF9SUUE6JY in limited sync - not resetting
Got remote echo for event $jFY2FyWh5xaypsaPLPYZ11AtDTa8zTPc5KF9SUUE6JY -> $jFY2FyWh5xaypsaPLPYZ11AtDTa8zTPc5KF9SUUE6JY old status sent
sendEvent of type m.room.message in !cJFtAIkwxuofiSYkPN:matrix.org with txnId m1611827663768.350
Event sent to !cJFtAIkwxuofiSYkPN:matrix.org with event id $6RimZRXPttjX-8K-V3KqKRLcwJlplMRJStM29fnulGQ
setting pendingEvent status to sent in !cJFtAIkwxuofiSYkPN:matrix.org event ID ~!cJFtAIkwxuofiSYkPN:matrix.org:m1611827663768.350 -> $6RimZRXPttjX-8K-V3KqKRLcwJlplMRJStM29fnulGQ
Already have event $6RimZRXPttjX-8K-V3KqKRLcwJlplMRJStM29fnulGQ in limited sync - not resetting
Got remote echo for event $6RimZRXPttjX-8K-V3KqKRLcwJlplMRJStM29fnulGQ -> $6RimZRXPttjX-8K-V3KqKRLcwJlplMRJStM29fnulGQ old status sent
~ » 

----

server log
[..]
[2021-01-28T09:42:33.564] [INFO] default - 
2021-01-28 09:43:33          API-WS: disconnected from wss://westend-rpc.polkadot.io/: 1006:: Connection dropped by remote peer.
[2021-01-28T09:43:34.705] [INFO] default - 
2021-01-28 09:44:34          API-WS: disconnected from wss://westend-rpc.polkadot.io/: 1006:: Connection dropped by remote peer.
[2021-01-28T09:44:35.804] [INFO] default - 
2021-01-28 09:45:35          API-WS: disconnected from wss://westend-rpc.polkadot.io/: 1006:: Connection dropped by remote peer.
[2021-01-28T09:45:36.964] [INFO] default - 
2021-01-28 09:46:36          API-WS: disconnected from wss://westend-rpc.polkadot.io/: 1006:: Connection dropped by remote peer.
[2021-01-28T09:46:38.056] [INFO] default - 
2021-01-28 09:47:38          API-WS: disconnected from wss://westend-rpc.polkadot.io/: 1006:: Connection dropped by remote peer.
[2021-01-28T09:47:39.170] [INFO] default - 
2021-01-28 09:48:39          API-WS: disconnected from wss://westend-rpc.polkadot.io/: 1006:: Connection dropped by remote peer.
[2021-01-28T09:48:40.304] [INFO] default - 
2021-01-28 09:49:40          API-WS: disconnected from wss://westend-rpc.polkadot.io/: 1006:: Connection dropped by remote peer.
[2021-01-28T09:49:41.423] [INFO] default - 
2021-01-28 09:50:41          API-WS: disconnected from wss://westend-rpc.polkadot.io/: 1006:: Connection dropped by remote peer.
[2021-01-28T09:50:42.536] [INFO] default - 
2021-01-28 09:51:42          API-WS: disconnected from wss://westend-rpc.polkadot.io/: 1006:: Connection dropped by remote peer.
[2021-01-28T09:51:43.676] [INFO] default - 
2021-01-28 09:52:43          API-WS: disconnected from wss://westend-rpc.polkadot.io/: 1006:: Connection dropped by remote peer.
[2021-01-28T09:52:44.773] [INFO] default - 
2021-01-28 09:53:44          API-WS: disconnected from wss://westend-rpc.polkadot.io/: 1006:: Connection dropped by remote peer.
[2021-01-28T09:53:45.893] [INFO] default - 
[2021-01-28T09:54:06.181] [ERROR] default - Oops, drip took more than 10000ms to answer
2021-01-28 09:54:45          API-WS: disconnected from wss://westend-rpc.polkadot.io/: 1006:: Connection dropped by remote peer.
[2021-01-28T09:54:46.995] [INFO] default - 
2021-01-28 09:55:46          API-WS: disconnected from wss://westend-rpc.polkadot.io/: 1006:: Connection dropped by remote peer.
[2021-01-28T09:55:48.109] [INFO] default - 
2021-01-28 09:56:48          API-WS: disconnected from wss://westend-rpc.polkadot.io/: 1006:: Connection dropped by remote peer.
[2021-01-28T09:56:49.230] [INFO] default - 
2021-01-28 09:57:49          API-WS: disconnected from wss://westend-rpc.polkadot.io/: 1006:: Connection dropped by remote peer.
[2021-01-28T09:57:50.361] [INFO] default - 

----

```

--- a/src/bot/index.ts
+++ b/src/bot/index.ts
@@ -62,7 +62,7 @@ bot.on('RoomMember.membership', (_, member: Record<string, string>) => {
   if (member.membership === 'invite' && member.userId === botUserId) {
     bot.joinRoom(member.roomId).then(() => {
       logger.info(`Auto-joined ${member.roomId}.`);
-    }).catch((e) => logger.error('Auto-join error', e));
+    }).catch((e) => logger.error('⭕ Auto-join error', e));
   }
 });
 
@@ -91,7 +91,7 @@ bot.on('Room.timeline', (event: mSDK.MatrixEvent) => {
       sendMessage(roomId, `The faucet has ${balance / 10 ** decimals} ${unit}s remaining.`);
     }).catch((e) => {
       sendMessage(roomId, 'An error occured, please check the server logs.');
-      logger.error('An error occured when checking the balance', e);
+      logger.error('⭕ An error occured when checking the balance', e);
     });
   } else if (action === '!drip') {
     try {
@@ -124,7 +124,7 @@ bot.on('Room.timeline', (event: mSDK.MatrixEvent) => {
       sendMessage(roomId, `Sent ${sender} ${dripAmount} ${unit}s. Extrinsic hash: ${res.data as string}`);
     }).catch((e) => {
       sendMessage(roomId, 'An unexpected error occured, please check the server logs');
-      logger.error('An error occured when dripping', e);
+      logger.error('⭕ An error occured when dripping', e);
     });
   } else if (action === '!help') {
     printHelpMessage(roomId);

--- a/src/server/actions.ts
+++ b/src/server/actions.ts
@@ -18,52 +18,65 @@ const rpcTimeout = (service: string) => {
   const timeout = 10000;
   return setTimeout(() => {
     // log an error in console and in prometheus if the timeout is reached
-    logger.error(`Oops, ${service} took more than ${timeout}ms to answer`);
+    logger.error(`⭕ Oops, ${service} took more than ${timeout}ms to answer`);
     errorCounter.plusOne('rpcTimeout');
   }, timeout);
 };
 
 export default class Actions {
-  api: ApiPromise | undefined;
+  // api: ApiPromise | undefined;
   account: KeyringPair | undefined;
 
   constructor () {
-    this.createApi().then(() => {
+    this.getConnectedApiInstance().then((api) => {
+      logger.info('🤖 Beep bop - Creating the bot\'s account');
+
       // once the api is initialized, we can create and account
       // if we don't wait we'll get an error "@polkadot/wasm-crypto has not been initialized"
       const keyring = new Keyring({ type: 'sr25519' });
       this.account = keyring.addFromMnemonic(mnemonic);
-    }).catch(e => {
+
+      this.disconnectApi(api);
+    }).catch((e) => {
       logger.error(e);
       errorCounter.plusOne('other');
     });
   }
 
-  async createApi (): Promise<void> {
-    const provider = new WsProvider(url);
-    try {
-      this.api = await ApiPromise.create({ provider, types: injectedTypes });
-      this.api.on('connected', () => {
-        logger.log('--> api reconnected!');
-      });
-    } catch (e) {
-      logger.error(e);
-      errorCounter.plusOne('other');
-    }
+  async getConnectedApiInstance (): Promise<ApiPromise> {
+    // do not autoconnect
+    const provider = new WsProvider(url, false);
+
+    const api = new ApiPromise({ provider, types: injectedTypes });
+
+    // commented out as to not spam, but useful for debugging
+    // to make sure we connect only once.
+
+    // api.on('connected', () => {
+    //   logger.info('--> api connected');
+    // });
+
+    // api.on('disconnected', () => {
+    //   logger.info('<-- api disconnected');
+    // });
+
+    await api.connect();
+    await api.isReady;
+    return api;
   }
 
   async sendTokens (address: string, amount: string): Promise<string | null> {
     try {
-      if (!this.api?.isReady) {
-        throw new Error('api not ready');
-      }
+      const api = await this.getConnectedApiInstance();
 
       if (!this.account) {
         throw new Error('account not ready');
       }
 
+      logger.info('💸 sending tokens');
+
       const dripAmount = Number(amount) * 10 ** decimals;
-      const transfer = this.api.tx.balances.transfer(address, dripAmount);
+      const transfer = api.tx.balances.transfer(address, dripAmount);
 
       // start a counter and log a timeout error if we didn't get an answer in time
       const dripTimeout = rpcTimeout('drip');
@@ -71,31 +84,44 @@ export default class Actions {
 
       // we got and answer reset the timeout
       clearTimeout(dripTimeout);
+
+      this.disconnectApi(api);
       return hash.toHex();
     } catch (e) {
-      logger.error('An error occured when sending tokens', e);
+      logger.error('⭕ An error occured when sending tokens', e);
       errorCounter.plusOne('other');
       return null;
     }
   }
 
   async getBalance (): Promise<string> {
-    if (!this.api?.isReady) {
-      throw new Error('api not ready');
+    try {
+      const api = await this.getConnectedApiInstance();
+
+      if (!this.account) {
+        throw new Error('account not ready');
+      }
+
+      logger.info('💰 checking balance');
+
+      // start a counter and log a timeout error if we didn't get an answer in time
+      const balanceTimeout = rpcTimeout('balance');
+
+      const { data: balances } = await api.query.system.account(this.account.address);
+
+      // we got and answer reset the timeout
+      clearTimeout(balanceTimeout);
+
+      this.disconnectApi(api);
+      return balances.free.toString();
+    } catch (e) {
+      logger.error('⭕ An error occured when querying the balance', e);
+      errorCounter.plusOne('other');
+      return '0';
     }
+  }
 
-    if (!this.account) {
-      throw new Error('account not ready');
-    }
-
-    // start a counter and log a timeout error if we didn't get an answer in time
-    const balanceTimeout = rpcTimeout('balance');
-
-    const { data: balances } = await this.api.query.system.account(this.account.address);
-
-    // we got and answer reset the timeout
-    clearTimeout(balanceTimeout);
-
-    return balances.free.toString();
+  disconnectApi (api: ApiPromise): void {
+    api.disconnect().catch((e) => logger.error('⭕ disconnection error', e));
   }
 }

